File: src/utils/encoding.ts
import fs from 'node:fs/promises';
import chardet from 'chardet';
import iconv from 'iconv-lite';
import { getNativeChunker } from '../chunking/nativeChunker.js';

/**
 * 支持的编码列表（按优先级排序）
 */
const _SUPPORTED_ENCODINGS = [
  'UTF-8',
  'UTF-16 LE',
  'UTF-16 BE',
  'UTF-32 LE',
  'UTF-32 BE',
  'GB18030', // 兼容 GBK 和 GB2312
  'Big5',
  'Shift_JIS',
  'EUC-JP',
  'EUC-KR',
  'ISO-8859-1',
  'windows-1252',
];

/**
 * 规范化编码名称，使其与 iconv-lite 兼容
 */
function normalizeEncoding(encoding: string): string {
  const map: Record<string, string> = {
    'UTF-8': 'utf8',
    'UTF-16 LE': 'utf16le',
    'UTF-16 BE': 'utf16be',
    'UTF-32 LE': 'utf32le',
    'UTF-32 BE': 'utf32be',
    GB18030: 'gb18030',
    GBK: 'gbk',
    GB2312: 'gb2312',
    Big5: 'big5',
    Shift_JIS: 'shiftjis',
    'EUC-JP': 'eucjp',
    'EUC-KR': 'euckr',
    'ISO-8859-1': 'iso88591',
    'windows-1252': 'win1252',
    ASCII: 'utf8', // ASCII 是 UTF-8 的子集
  };
  return map[encoding] || encoding.toLowerCase().replace(/[^a-z0-9]/g, '');
}

/**
 * 检测 BOM（Byte Order Mark）
 */
function detectBOM(buffer: Buffer): string | null {
  if (buffer.length >= 3) {
    // UTF-8 BOM
    if (buffer[0] === 0xef && buffer[1] === 0xbb && buffer[2] === 0xbf) {
      return 'UTF-8';
    }
  }
  if (buffer.length >= 4) {
    // UTF-32 LE
    if (buffer[0] === 0xff && buffer[1] === 0xfe && buffer[2] === 0x00 && buffer[3] === 0x00) {
      return 'UTF-32 LE';
    }
    // UTF-32 BE
    if (buffer[0] === 0x00 && buffer[1] === 0x00 && buffer[2] === 0xfe && buffer[3] === 0xff) {
      return 'UTF-32 BE';
    }
  }
  if (buffer.length >= 2) {
    // UTF-16 LE
    if (buffer[0] === 0xff && buffer[1] === 0xfe) {
      return 'UTF-16 LE';
    }
    // UTF-16 BE
    if (buffer[0] === 0xfe && buffer[1] === 0xff) {
      return 'UTF-16 BE';
    }
  }
  return null;
}

/**
 * 读取文件并自动转换编码为 UTF-8
 * @param filePath 文件路径
 * @returns 解码后的文件内容和检测到的编码
 */
export async function readFileWithEncoding(filePath: string): Promise<{
  content: string;
  encoding: string;
  originalEncoding: string;
}> {
  const buffer = await fs.readFile(filePath);

  // 优先用 Rust native 检测+解码（chardetng + encoding_rs）；不可用或抛错时回退 TS。
  const native = getNativeChunker();
  if (native) {
    try {
      const { content, originalEncoding } = native.decodeBytes(buffer);
      return { content, encoding: 'utf-8', originalEncoding };
    } catch {
      // 回退到 TS chardet/iconv 路径
    }
  }

  return decodeBytesTs(buffer);
}

/**
 * TS 解码路径（chardet 检测 + iconv-lite 转码），native 不可用时的 fallback。
 */
function decodeBytesTs(buffer: Buffer): {
  content: string;
  encoding: string;
  originalEncoding: string;
} {
  // 检测编码
  const bom = detectBOM(buffer);
  let encoding = bom;

  if (!encoding) {
    const detected = chardet.detect(buffer);
    encoding = detected || 'UTF-8';
  }

  const normalizedEncoding = normalizeEncoding(encoding);

  // 尝试使用检测到的编码解码
  let content: string;
  try {
    if (iconv.encodingExists(normalizedEncoding)) {
      content = iconv.decode(buffer, normalizedEncoding);
    } else {
      // 回退到 UTF-8
      content = buffer.toString('utf-8');
    }
  } catch {
    // 解码失败，回退到 UTF-8
    content = buffer.toString('utf-8');
  }

  return {
    content,
    encoding: 'utf-8', // 输出始终是 UTF-8
    originalEncoding: encoding,
  };
}

File: crates/chunker/src/encoding.rs
//! Port of `src/utils/encoding.ts` `readFileWithEncoding()` decode step.
//!
//! Takes raw file bytes and returns UTF-8 text plus the detected source
//! encoding label. Mirrors the TS pipeline: BOM sniff → chardet detection →
//! decode. The TS side keeps the `fs.readFile` (I/O stays in Node); only the
//! CPU-bound detect+decode is moved here.
//!
//! Detector: `chardetng` (Firefox's encoding detector) replaces the JS
//! `chardet` crate; decoder: `encoding_rs` (Gecko engine) replaces `iconv-lite`.
//! Output is always valid UTF-8 (lossy replacement on malformed input, matching
//! iconv-lite's default tolerant behavior).

use encoding_rs::Encoding;

/// Detect a BOM at the start of `buffer`, returning the matching encoding plus
/// the BOM length to strip. Order mirrors `detectBOM` in encoding.ts: UTF-8,
/// then UTF-32 (must precede UTF-16 since UTF-32 LE begins with the UTF-16 LE
/// BOM bytes), then UTF-16.
fn detect_bom(buffer: &[u8]) -> Option<(&'static Encoding, usize)> {
    if buffer.len() >= 3 && buffer[0] == 0xef && buffer[1] == 0xbb && buffer[2] == 0xbf {
        return Some((encoding_rs::UTF_8, 3));
    }
    if buffer.len() >= 4 {
        if buffer[0] == 0xff && buffer[1] == 0xfe && buffer[2] == 0x00 && buffer[3] == 0x00 {
            // UTF-32 LE — encoding_rs has no UTF-32; handled specially by caller.
            return Some((encoding_rs::UTF_16LE, 4)); // sentinel; see decode()
        }
        if buffer[0] == 0x00 && buffer[1] == 0x00 && buffer[2] == 0xfe && buffer[3] == 0xff {
            return Some((encoding_rs::UTF_16BE, 4)); // sentinel; see decode()
        }
    }
    if buffer.len() >= 2 {
        if buffer[0] == 0xff && buffer[1] == 0xfe {
            return Some((encoding_rs::UTF_16LE, 2));
        }
        if buffer[0] == 0xfe && buffer[1] == 0xff {
            return Some((encoding_rs::UTF_16BE, 2));
        }
    }
    None
}

/// Decode UTF-32 (LE/BE) by hand — `encoding_rs` does not support it but
/// iconv-lite does, so the TS path can produce UTF-32 content we must match.
/// `le` selects byte order; the 4-byte BOM is assumed already stripped.
fn decode_utf32(bytes: &[u8], le: bool) -> String {
    let mut out = String::with_capacity(bytes.len() / 4);
    for chunk in bytes.chunks_exact(4) {
        let cp = if le {
            u32::from_le_bytes([chunk[0], chunk[1], chunk[2], chunk[3]])
        } else {
            u32::from_be_bytes([chunk[0], chunk[1], chunk[2], chunk[3]])
        };
        out.push(char::from_u32(cp).unwrap_or('\u{fffd}'));
    }
    out
}

/// Result of decoding raw bytes to UTF-8.
pub struct Decoded {
    pub content: String,
    pub original_encoding: String,
}

/// Detect encoding and decode `buffer` to UTF-8.
///
/// 1. BOM sniff (authoritative when present).
/// 2. Otherwise run chardetng across the whole buffer.
/// 3. Decode with the resolved encoding; UTF-32 handled manually.
pub fn decode_bytes(buffer: &[u8]) -> Decoded {
    // 1. BOM
    if let Some((enc, bom_len)) = detect_bom(buffer) {
        let body = &buffer[bom_len..];
        // UTF-32 sentinels: 4-byte BOM with the LE/BE markers handled here.
        if bom_len == 4 {
            let le = buffer[0] == 0xff;
            return Decoded {
                content: decode_utf32(body, le),
                original_encoding: if le { "UTF-32 LE".into() } else { "UTF-32 BE".into() },
            };
        }
        let (cow, _, _) = enc.decode(body);
        return Decoded {
            content: cow.into_owned(),
            original_encoding: bom_label(enc),
        };
    }

    // 2. chardetng detection over the full buffer.
    let mut detector = chardetng::EncodingDetector::new();
    detector.feed(buffer, true);
    let enc = detector.guess(None, true);

    // 3. decode
    let (cow, _, _) = enc.decode(buffer);
    Decoded {
        content: cow.into_owned(),
        original_encoding: enc.name().to_string(),
    }
}

/// Map BOM-resolved encoding to the label TS reports in `originalEncoding`.
fn bom_label(enc: &'static Encoding) -> String {
    if enc == encoding_rs::UTF_8 {
        "UTF-8".into()
    } else if enc == encoding_rs::UTF_16LE {
        "UTF-16 LE".into()
    } else if enc == encoding_rs::UTF_16BE {
        "UTF-16 BE".into()
    } else {
        enc.name().to_string()
    }
}
